diff --git a/ext/libav/gstavcodecmap.c b/ext/libav/gstavcodecmap.c
index df3aa6b..1d40153 100644
--- a/ext/libav/gstavcodecmap.c
+++ b/ext/libav/gstavcodecmap.c
@@ -3520,14 +3520,16 @@ gst_ffmpeg_formatid_to_caps (const gchar * format_name)
   GstCaps *caps = NULL;
 
   if (!strcmp (format_name, "mpeg")) {
-    caps = gst_caps_new_simple ("video/mpeg",
-        "systemstream", G_TYPE_BOOLEAN, TRUE, NULL);
+    //caps = gst_caps_new_simple ("video/mpeg",
+    //    "systemstream", G_TYPE_BOOLEAN, TRUE, NULL);
+    caps = gst_caps_from_string ("video/x-cdxa");
   } else if (!strcmp (format_name, "mpegts")) {
     caps = gst_caps_new_simple ("video/mpegts",
         "systemstream", G_TYPE_BOOLEAN, TRUE, NULL);
   } else if (!strcmp (format_name, "rm")) {
-    caps = gst_caps_new_simple ("application/x-pn-realmedia",
-        "systemstream", G_TYPE_BOOLEAN, TRUE, NULL);
+    //caps = gst_caps_new_simple ("application/x-pn-realmedia",
+    //    "systemstream", G_TYPE_BOOLEAN, TRUE, NULL);
+    caps = gst_caps_new_empty_simple ("application/vnd.rn-realmedia");
   } else if (!strcmp (format_name, "asf")) {
     caps = gst_caps_new_empty_simple ("video/x-ms-asf");
   } else if (!strcmp (format_name, "avi")) {
@@ -3545,7 +3547,7 @@ gst_ffmpeg_formatid_to_caps (const gchar * format_name)
         "systemstream", G_TYPE_BOOLEAN, TRUE, NULL);
   } else if (!strcmp (format_name, "4xm")) {
     caps = gst_caps_new_empty_simple ("video/x-4xm");
-  } else if (!strcmp (format_name, "matroska")) {
+  } else if (!strcmp (format_name, "matroska_webm")) {
     caps = gst_caps_new_empty_simple ("video/x-matroska");
   } else if (!strcmp (format_name, "ivf")) {
     caps = gst_caps_new_empty_simple ("video/x-ivf");
@@ -3559,7 +3561,7 @@ gst_ffmpeg_formatid_to_caps (const gchar * format_name)
     caps = gst_caps_new_empty_simple ("audio/x-ttafile");
   } else if (!strcmp (format_name, "aiff")) {
     caps = gst_caps_new_empty_simple ("audio/x-aiff");
-  } else if (!strcmp (format_name, "mov_mp4_m4a_3gp_3g2")) {
+  } else if (!strcmp (format_name, "mov_mp4_m4a_3gp_3g2_mj2")) {
     caps =
         gst_caps_from_string
         ("application/x-3gp; video/quicktime; audio/x-m4a");
@@ -3581,7 +3583,8 @@ gst_ffmpeg_formatid_to_caps (const gchar * format_name)
   } else if (!strcmp (format_name, "gif")) {
     caps = gst_caps_from_string ("image/gif");
   } else if (!strcmp (format_name, "ogg")) {
-    caps = gst_caps_from_string ("application/ogg");
+    //caps = gst_caps_from_string ("application/ogg");
+    caps = gst_caps_from_string ("video/ogg");
   } else if (!strcmp (format_name, "mxf") || !strcmp (format_name, "mxf_d10")) {
     caps = gst_caps_from_string ("application/mxf");
   } else if (!strcmp (format_name, "gxf")) {
diff --git a/ext/libav/gstavdemux.c b/ext/libav/gstavdemux.c
index 37de3fa..0fd47bf 100644
--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -2015,6 +2015,7 @@ gst_ffmpegdemux_register (GstPlugin * plugin)
   };
 
   void *i = 0;
+  const char *env;
 
   GST_LOG ("Registering demuxers");
 
@@ -2104,58 +2105,63 @@ gst_ffmpegdemux_register (GstPlugin * plugin)
         !strcmp (in_plugin->name, "ivf"))
       register_typefind_func = FALSE;
 
-    /* Set the rank of demuxers known to work to MARGINAL.
-     * Set demuxers for which we already have another implementation to NONE
-     * Set All others to NONE*/
-    /**
-     * element-avdemux_xwma
-     *
-     * Since: 1.20
-     */
-    if (!strcmp (in_plugin->name, "wsvqa") ||
-        !strcmp (in_plugin->name, "wsaud") ||
-        !strcmp (in_plugin->name, "wc3movie") ||
-        !strcmp (in_plugin->name, "voc") ||
-        !strcmp (in_plugin->name, "tta") ||
-        !strcmp (in_plugin->name, "sol") ||
-        !strcmp (in_plugin->name, "smk") ||
-        !strcmp (in_plugin->name, "vmd") ||
-        !strcmp (in_plugin->name, "film_cpk") ||
-        !strcmp (in_plugin->name, "ingenient") ||
-        !strcmp (in_plugin->name, "psxstr") ||
-        !strcmp (in_plugin->name, "nuv") ||
-        !strcmp (in_plugin->name, "nut") ||
-        !strcmp (in_plugin->name, "nsv") ||
-        !strcmp (in_plugin->name, "mxf") ||
-        !strcmp (in_plugin->name, "mmf") ||
-        !strcmp (in_plugin->name, "mm") ||
-        !strcmp (in_plugin->name, "ipmovie") ||
-        !strcmp (in_plugin->name, "ape") ||
-        !strcmp (in_plugin->name, "RoQ") ||
-        !strcmp (in_plugin->name, "idcin") ||
-        !strcmp (in_plugin->name, "gxf") ||
-        !strcmp (in_plugin->name, "ffm") ||
-        !strcmp (in_plugin->name, "ea") ||
-        !strcmp (in_plugin->name, "daud") ||
-        !strcmp (in_plugin->name, "avs") ||
-        !strcmp (in_plugin->name, "aiff") ||
-        !strcmp (in_plugin->name, "xwma") ||
-        !strcmp (in_plugin->name, "4xm") ||
-        !strcmp (in_plugin->name, "yuv4mpegpipe") ||
-        !strcmp (in_plugin->name, "pva") ||
-        !strcmp (in_plugin->name, "mpc") ||
-        !strcmp (in_plugin->name, "mpc8") ||
-        !strcmp (in_plugin->name, "ivf") ||
-        !strcmp (in_plugin->name, "brstm") ||
-        !strcmp (in_plugin->name, "bfstm") ||
-        !strcmp (in_plugin->name, "gif") ||
-        !strcmp (in_plugin->name, "dsf") || !strcmp (in_plugin->name, "iff"))
-      rank = GST_RANK_MARGINAL;
-    else {
-      GST_DEBUG ("ignoring %s", in_plugin->name);
-      rank = GST_RANK_NONE;
-      continue;
-    }
+     env = getenv("VENDOR_MEDIA_GST_LIBAV_ENABLE_FFMPEG");
+     if ((NULL == env) || ((env) && (atoi(env)))) {
+         rank = GST_RANK_PRIMARY;
+     } else {
+         /* Set the rank of demuxers known to work to MARGINAL.
+          * Set demuxers for which we already have another implementation to NONE
+            Set All others to NONE*/
+         /**
+          * element-avdemux_xwma
+          *
+          * Since: 1.20
+          */
+        if (!strcmp (in_plugin->name, "wsvqa") ||
+            !strcmp (in_plugin->name, "wsaud") ||
+            !strcmp (in_plugin->name, "wc3movie") ||
+            !strcmp (in_plugin->name, "voc") ||
+            !strcmp (in_plugin->name, "tta") ||
+            !strcmp (in_plugin->name, "sol") ||
+            !strcmp (in_plugin->name, "smk") ||
+            !strcmp (in_plugin->name, "vmd") ||
+            !strcmp (in_plugin->name, "film_cpk") ||
+            !strcmp (in_plugin->name, "ingenient") ||
+            !strcmp (in_plugin->name, "psxstr") ||
+            !strcmp (in_plugin->name, "nuv") ||
+            !strcmp (in_plugin->name, "nut") ||
+            !strcmp (in_plugin->name, "nsv") ||
+            !strcmp (in_plugin->name, "mxf") ||
+            !strcmp (in_plugin->name, "mmf") ||
+            !strcmp (in_plugin->name, "mm") ||
+            !strcmp (in_plugin->name, "ipmovie") ||
+            !strcmp (in_plugin->name, "ape") ||
+            !strcmp (in_plugin->name, "RoQ") ||
+            !strcmp (in_plugin->name, "idcin") ||
+            !strcmp (in_plugin->name, "gxf") ||
+            !strcmp (in_plugin->name, "ffm") ||
+            !strcmp (in_plugin->name, "ea") ||
+            !strcmp (in_plugin->name, "daud") ||
+            !strcmp (in_plugin->name, "avs") ||
+            !strcmp (in_plugin->name, "aiff") ||
+            !strcmp (in_plugin->name, "xwma") ||
+            !strcmp (in_plugin->name, "4xm") ||
+            !strcmp (in_plugin->name, "yuv4mpegpipe") ||
+            !strcmp (in_plugin->name, "pva") ||
+            !strcmp (in_plugin->name, "mpc") ||
+            !strcmp (in_plugin->name, "mpc8") ||
+            !strcmp (in_plugin->name, "ivf") ||
+            !strcmp (in_plugin->name, "brstm") ||
+            !strcmp (in_plugin->name, "bfstm") ||
+            !strcmp (in_plugin->name, "gif") ||
+            !strcmp (in_plugin->name, "dsf") || !strcmp (in_plugin->name, "iff"))
+          rank = GST_RANK_MARGINAL;
+        else {
+          GST_DEBUG ("ignoring %s", in_plugin->name);
+          rank = GST_RANK_NONE;
+          continue;
+        }
+     }
 
     /* construct the type */
     type_name = g_strdup_printf ("avdemux_%s", in_plugin->name);
