diff --git a/ext/libav/gstavdemux.c b/ext/libav/gstavdemux.c
index 0fd47bf..daf50b2 100644
--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -36,7 +36,7 @@
 #include "gstavutils.h"
 #include "gstavprotocol.h"
 
-#define MAX_STREAMS 20
+#define MAX_STREAMS 40
 
 typedef struct _GstFFMpegDemux GstFFMpegDemux;
 typedef struct _GstFFStream GstFFStream;
@@ -1279,6 +1279,11 @@ gst_ffmpegdemux_open (GstFFMpegDemux * demux)
 
   n_streams = demux->context->nb_streams;
   GST_DEBUG_OBJECT (demux, "we have %d streams", n_streams);
+  if (n_streams > MAX_STREAMS)
+  {
+    GST_ERROR_OBJECT (demux, "The n_streams %d exceeds the MAX_STREAMS(40) ", n_streams);
+    goto beach;
+  }
 
   /* open_input_file() automatically reads the header. We can now map each
    * created AVStream to a GstPad to make GStreamer handle it. */
