diff --git a/ext/libav/gstavcodecmap.c b/ext/libav/gstavcodecmap.c
index 732f0c4..6db76f0 100644
--- a/ext/libav/gstavcodecmap.c
+++ b/ext/libav/gstavcodecmap.c
@@ -2295,6 +2295,11 @@ gst_ffmpeg_codecid_to_caps (enum AVCodecID codec_id,
       gst_caps_set_simple (caps, "lsbf", G_TYPE_BOOLEAN,
           FALSE, "planar", G_TYPE_BOOLEAN, TRUE, NULL);
       break;
+    case AV_CODEC_ID_AC4:
+      caps =
+          gst_ff_aud_caps_new (context, NULL, codec_id, encode, "audio/x-ac4",
+          NULL);
+      break;
     default:
       GST_DEBUG ("Unknown codec ID %d, please add mapping here", codec_id);
       break;
