diff --git a/ext/libav/gstavcodecmap.c b/ext/libav/gstavcodecmap.c
index 1d40153..4592cdf 100644
--- a/ext/libav/gstavcodecmap.c
+++ b/ext/libav/gstavcodecmap.c
@@ -2349,6 +2349,11 @@ gst_ffmpeg_codecid_to_caps (enum AVCodecID codec_id,
           gst_ff_vid_caps_new (context, NULL, codec_id, encode, "video/x-av1",
           NULL);
       break;
+    case AV_CODEC_ID_AC4:
+      caps =
+          gst_ff_aud_caps_new (context, NULL, codec_id, encode, "audio/x-ac4",
+          NULL);
+      break;
     default:
       GST_DEBUG ("Unknown codec ID %d, please add mapping here", codec_id);
       break;
