diff --git a/ext/libav/gstavdemux.c b/ext/libav/gstavdemux.c
old mode 100644
new mode 100755
index e45925c..7fbb339
--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -2089,7 +2089,10 @@ gst_ffmpegdemux_register (GstPlugin * plugin)
 
     env = getenv("ENABLE_FFMPEG_DEMUX");
     if ((env) && (!strcmp(env, "TRUE"))) {
-        rank = GST_RANK_PRIMARY;
+		if(!strcmp(in_plugin->name, "mpegts"))
+			rank = GST_RANK_NONE;
+		else
+            rank = GST_RANK_PRIMARY;
     } else {
         /* Set the rank of demuxers known to work to MARGINAL.
          * Set demuxers for which we already have another implementation to NONE
