diff --git a/ext/libav/gstavdemux.c b/ext/libav/gstavdemux.c
index daf50b2..621c975 100644
--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -1482,11 +1482,11 @@ gst_ffmpegdemux_loop (GstFFMpegDemux * demux)
      * -1
      * 1
      *
-     * we reset pts to 0 since for us timestamp are unsigned
+     * we reset pts to AV_NOPTS_VALUE since for us timestamp are unsigned
      */
     GST_WARNING_OBJECT (demux,
-        "negative pts detected: %" G_GINT64_FORMAT " resetting to 0", pts);
-    pts = 0;
+        "negative pts detected: %" G_GINT64_FORMAT " resetting to AV_NOPTS_VALUE", pts);
+    pts = AV_NOPTS_VALUE;
   }
   timestamp = gst_ffmpeg_time_ff_to_gst (pts, avstream->time_base);
   if (GST_CLOCK_TIME_IS_VALID (timestamp)) {
