diff --git a/ext/libav/gstavdemux.c b/ext/libav/gstavdemux.c
index 621c975..39bcfd9 100644
--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -972,6 +972,14 @@ gst_ffmpegdemux_get_stream (GstFFMpegDemux * demux, AVStream * avstream)
   caps = gst_ffmpeg_codecid_to_caps (ctx->codec_id, ctx, TRUE);
   if (caps == NULL)
     goto unknown_caps;
+  else if ((!strcmp(demux->context->iformat->name, "mpegts")
+       && (AV_CODEC_ID_H264 == ctx->codec_id || AV_CODEC_ID_HEVC == ctx->codec_id))
+       || (!strcmp(demux->context->iformat->name, "avi")
+       && (AV_CODEC_ID_H264 == ctx->codec_id || AV_CODEC_ID_HEVC == ctx->codec_id)))
+      {
+         gst_caps_set_simple (caps, "parsed", G_TYPE_BOOLEAN, "true",
+             "stream-format", G_TYPE_STRING, "byte-stream", NULL);
+      }
 
   /* stream is known now */
   stream->unknown = FALSE;
