diff --git a/ext/libav/gstavdemux.c b/ext/libav/gstavdemux.c
index 39bcfd9..97e0296 100644
--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -1301,6 +1301,11 @@ gst_ffmpegdemux_open (GstFFMpegDemux * demux)
 
   gst_element_no_more_pads (GST_ELEMENT (demux));
 
+  if(0 == strcmp(demux->context->iformat->name, "wav") && AV_NOPTS_VALUE == demux->context->start_time){
+    /* start time is nops in wav file, set it to 0 */
+    GST_DEBUG_OBJECT (demux, "start time is nops in wav file, set it to 0.");
+    demux->context->start_time=0;
+  }
   /* transform some useful info to GstClockTime and remember */
   demux->start_time = gst_util_uint64_scale_int (demux->context->start_time,
       GST_SECOND, AV_TIME_BASE);
