diff --git a/ext/libav/gstavdemux.c b/ext/libav/gstavdemux.c
index 97e0296..72e89ea 100644
--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -1293,6 +1293,15 @@ gst_ffmpegdemux_open (GstFFMpegDemux * demux)
     goto beach;
   }
 
+  //check dolbyvision config box
+  for (i = 0; i < n_streams; i++) {
+    if (demux->context->streams[i]->codec->codec_type == AVMEDIA_TYPE_VIDEO &&
+         demux->context->streams[i]->codec->has_dolby_vision_config_box == AV_DV_BOX_TYPE_ERROR) {
+         GST_ERROR_OBJECT (demux, "Not support dolbyvison config box");
+         goto beach;
+    }
+  }
+
   /* open_input_file() automatically reads the header. We can now map each
    * created AVStream to a GstPad to make GStreamer handle it. */
   for (i = 0; i < n_streams; i++) {
