diff --git a/ext/libav/gstavdemux.c b/ext/libav/gstavdemux.c
index 45ddc0b..7951214 100644
--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -2136,7 +2136,8 @@ gst_ffmpegdemux_register (GstPlugin * plugin)
       register_typefind_func = FALSE;
 
      env = getenv("VENDOR_MEDIA_GST_LIBAV_ENABLE_FFMPEG");
-     if ((NULL == env) || ((env) && (atoi(env)))) {
+     if (((NULL == env) || ((env) && (atoi(env))))
+        && strcmp (in_plugin->name, "mov,mp4,m4a,3gp,3g2,mj2")){
          rank = GST_RANK_PRIMARY;
      } else {
          /* Set the rank of demuxers known to work to MARGINAL.
