diff --git a/ext/libav/gstavdemux.c b/ext/libav/gstavdemux.c
index 24ddfd6..3e82f97 100644
--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -2187,7 +2187,8 @@ gst_ffmpegdemux_register (GstPlugin * plugin)
           && strcmp (in_plugin->name, "mov,mp4,m4a,3gp,3g2,mj2")
           && strcmp (in_plugin->name, "matroska")
           && strcmp (in_plugin->name, "matroska_webm")
-          && strcmp (in_plugin->name, "matroska,webm")) {
+          && strcmp (in_plugin->name, "matroska,webm")
+          && strcmp (in_plugin->name, "mpegts")) {
          rank = GST_RANK_PRIMARY;
      } else {
          /* Set the rank of demuxers known to work to MARGINAL.
