diff --git a/ext/libav/gstavdemux.c b/ext/libav/gstavdemux.c
index 3e82f97..ff29b1c 100644
--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -2188,7 +2188,8 @@ gst_ffmpegdemux_register (GstPlugin * plugin)
           && strcmp (in_plugin->name, "matroska")
           && strcmp (in_plugin->name, "matroska_webm")
           && strcmp (in_plugin->name, "matroska,webm")
-          && strcmp (in_plugin->name, "mpegts")) {
+          && strcmp (in_plugin->name, "mpegts")
+          && strcmp (in_plugin->name, "wav")) {
          rank = GST_RANK_PRIMARY;
      } else {
          /* Set the rank of demuxers known to work to MARGINAL.
