diff --git a/ext/libav/gstavdemux.c b/ext/libav/gstavdemux.c
index ff29b1c..ee16595 100644
--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -2153,11 +2153,11 @@ gst_ffmpegdemux_register (GstPlugin * plugin)
         !strcmp (in_plugin->name, "avi") ||
         !strcmp (in_plugin->name, "asf") ||
         !strcmp (in_plugin->name, "mpegvideo") ||
-        !strcmp (in_plugin->name, "mp3") ||
+        //!strcmp (in_plugin->name, "mp3") ||
         !strcmp (in_plugin->name, "matroska") ||
         !strcmp (in_plugin->name, "matroska_webm") ||
         !strcmp (in_plugin->name, "matroska,webm") ||
-        !strcmp (in_plugin->name, "mpeg") ||
+        //!strcmp (in_plugin->name, "mpeg") ||
         !strcmp (in_plugin->name, "wav") ||
         !strcmp (in_plugin->name, "au") ||
         !strcmp (in_plugin->name, "tta") ||
