diff --git a/ext/libav/gstavviddec.c b/ext/libav/gstavviddec.c
index f8d77d9..030aeff 100644
--- a/ext/libav/gstavviddec.c
+++ b/ext/libav/gstavviddec.c
@@ -2569,6 +2569,7 @@ gst_ffmpegviddec_register (GstPlugin * plugin)
         !strcmp (in_plugin->name, "pgssub") ||
         !strcmp (in_plugin->name, "dvdsub") ||
         !strcmp (in_plugin->name, "dvbsub") ||
+        !strcmp (in_plugin->name, "vc1") ||
         !strcmp (in_plugin->name, "cdgraphics") ||
         !strcmp (in_plugin->name, "av1")) {
       GST_LOG ("Ignoring decoder %s", in_plugin->name);
