diff --git a/sys/v4l2/gstv4l2videodec.c b/sys/v4l2/gstv4l2videodec.c
index 3042995..8f0b00e 100755
--- a/sys/v4l2/gstv4l2videodec.c
+++ b/sys/v4l2/gstv4l2videodec.c
@@ -715,8 +715,8 @@ gst_v4l2_video_dec_handle_frame (GstVideoDecoder * decoder,
           GST_V4L2_MIN_BUFFERS (self->v4l2output));
       guint max = VIDEO_MAX_FRAME;
 
-      gst_buffer_pool_config_set_params (config, self->input_state->caps,
-          self->v4l2output->info.size, min, max);
+//      gst_buffer_pool_config_set_params (config, self->input_state->caps,
+//          self->v4l2output->info.size, min, max);
 
       /* There is no reason to refuse this config */
       if (!gst_buffer_pool_set_config (pool, config)) {
