diff --git a/sys/v4l2/gstv4l2videodec.c b/sys/v4l2/gstv4l2videodec.c
index 7969a93..b15e60d 100755
--- a/sys/v4l2/gstv4l2videodec.c
+++ b/sys/v4l2/gstv4l2videodec.c
@@ -1252,6 +1252,7 @@ gst_v4l2_video_dec_set_metadata (GstStructure * s, GstV4l2VideoDecCData * cdata,
 {
   gchar *codec_name = NULL;
   gchar *type_name = NULL;
+  gboolean got_value = FALSE;
 
 #define SET_META(codec) \
 G_STMT_START { \
@@ -1264,11 +1265,15 @@ G_STMT_START { \
     SET_META ("JPEG");
   } else if (gst_structure_has_name (s, "video/mpeg")) {
     gint mpegversion = 0;
-    gst_structure_get_int (s, "mpegversion", &mpegversion);
-
-    if (mpegversion == 2) {
-      SET_META ("MPEG2");
-      cdata->codec = gst_v4l2_mpeg2_get_codec ();
+    gint *list = NULL;
+    got_value = gst_structure_get_int (s, "mpegversion", &mpegversion);
+    if (FALSE == got_value) {
+      got_value = gst_structure_get_list (s, "mpegversion", &list);
+      if (TRUE == got_value && (1 == *list || 2 == *list)) {
+        SET_META ("MPEG2");
+      } else {
+        SET_META ("MPEG4");
+      }
     } else {
       SET_META ("MPEG4");
       cdata->codec = gst_v4l2_mpeg4_get_codec ();
