diff --git a/gst/isomp4/qtdemux.c b/gst/isomp4/qtdemux.c
index 24c489d..02764ba 100755
--- a/gst/isomp4/qtdemux.c
+++ b/gst/isomp4/qtdemux.c
@@ -3286,11 +3286,11 @@ qtdemux_parse_trun (GstQTDemux * qtdemux, GstByteReader * trun,
       GST_DEBUG_OBJECT (qtdemux,
           "invalid flags; SAMPLE and FIRST_SAMPLE present, discarding latter");
       flags ^= TR_FIRST_SAMPLE_FLAGS;
-    } else {
-      if (!gst_byte_reader_get_uint32_be (trun, &first_flags))
-        goto fail;
-      GST_LOG_OBJECT (qtdemux, "first flags: 0x%x", first_flags);
     }
+
+    if (!gst_byte_reader_get_uint32_be (trun, &first_flags))
+      goto fail;
+    GST_LOG_OBJECT (qtdemux, "first flags: 0x%x", first_flags);
   }
 
   /* FIXME ? spec says other bits should also be checked to determine
