diff --git a/gstv4l2object.c b/gstv4l2object.c
index 76295f2..98791b6 100644
--- a/gstv4l2object.c
+++ b/gstv4l2object.c
@@ -1449,6 +1449,8 @@ gst_v4l2_object_v4l2fourcc_to_bare_struct (guint32 fourcc)
     case V4L2_PIX_FMT_MPEG2:
       structure = gst_structure_new ("video/mpeg",
           "mpegversion", G_TYPE_INT, 2, NULL);
+      gst_structure_set (structure, "systemstream", G_TYPE_BOOLEAN, FALSE, NULL);
+      GST_DEBUG ("aml set mpeg2 systemstream to false");
       break;
     case V4L2_PIX_FMT_MPEG4:
     case V4L2_PIX_FMT_XVID:
