diff --git a/gst/isomp4/qtdemux.c b/gst/isomp4/qtdemux.c
index dea7b51..b324439 100755
--- a/gst/isomp4/qtdemux.c
+++ b/gst/isomp4/qtdemux.c
@@ -9540,11 +9540,13 @@ qtdemux_stbl_init (GstQTDemux * qtdemux, QtDemuxStream * stream, GNode * stbl)
           goto done;
         }
 
+#if 0
         /* Don't consider "no decode samples" with offset G_MININT32
          * for the DTS/PTS shift */
         if (offset != G_MININT32 && offset < cslg_least)
           cslg_least = offset;
+#endif
       }
 
       if (cslg_least < 0)
         stream->cslg_shift = -cslg_least;
