diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index e31b502..888304c 100755
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -1478,6 +1478,8 @@ gst_v4l2_object_v4l2fourcc_to_bare_struct (guint32 fourcc)
     case V4L2_PIX_FMT_MPEG2:
       structure = gst_structure_new ("video/mpeg",
           "mpegversion", G_TYPE_INT, 2, NULL);
+      gst_structure_set (structure, "systemstream", G_TYPE_BOOLEAN, FALSE, NULL);
+      GST_DEBUG ("aml set mpeg2 systemstream to false");
       break;
     case V4L2_PIX_FMT_MPEG4:
     case V4L2_PIX_FMT_XVID:
