diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index 69fa1fd..a90771a 100755
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -1491,6 +1491,8 @@ gst_v4l2_object_v4l2fourcc_to_bare_struct (guint32 fourcc)
     case V4L2_PIX_FMT_MPEG1:
       structure = gst_structure_new ("video/mpeg",
           "mpegversion", G_TYPE_INT, 1, NULL);
+      gst_structure_set (structure, "systemstream", G_TYPE_BOOLEAN, FALSE, NULL);
+      GST_DEBUG ("set mpeg1 systemstream to false");
       break;
     case V4L2_PIX_FMT_MPEG2:
       structure = gst_structure_new ("video/mpeg",
diff --git a/sys/v4l2/gstv4l2videodec.c b/sys/v4l2/gstv4l2videodec.c
index e0493ec..128a451 100755
--- a/sys/v4l2/gstv4l2videodec.c
+++ b/sys/v4l2/gstv4l2videodec.c
@@ -1394,7 +1394,7 @@ gst_v4l2_video_dec_register (GstPlugin * plugin, const gchar * basename,
     type_info.instance_init = gst_v4l2_video_dec_subinstance_init;
 
     subtype = g_type_register_static (type, type_name, &type_info, 0);
-    if (!gst_element_register (plugin, type_name, GST_RANK_PRIMARY + 1,
+    if (!gst_element_register (plugin, type_name, GST_RANK_NONE,
             subtype))
       GST_WARNING ("Failed to register plugin '%s'", type_name);
 
