diff --git a/gst/matroska/matroska-demux.c b/gst/matroska/matroska-demux.c
index f926823..f1fc616 100755
--- a/gst/matroska/matroska-demux.c
+++ b/gst/matroska/matroska-demux.c
@@ -6632,7 +6632,8 @@ gst_matroska_demux_video_caps (GstMatroskaTrackVideoContext *
     if (data) {
       GstBuffer *priv;
 
-      priv = gst_buffer_new_memdup (data, size);
+      //For now, only OBUs here, so skip 4 bytes OBU header
+      priv = gst_buffer_new_memdup (data+ 4, size-4);
       gst_caps_set_simple (caps, "codec_data", GST_TYPE_BUFFER, priv, NULL);
       gst_buffer_unref (priv);
     } else {
