diff --git a/gst/isomp4/fourcc.h b/gst/isomp4/fourcc.h
index 9a0be6b..2e75b11 100755
--- a/gst/isomp4/fourcc.h
+++ b/gst/isomp4/fourcc.h
@@ -408,6 +408,7 @@ G_BEGIN_DECLS
 #define FOURCC_frma     GST_MAKE_FOURCC('f','r','m','a')
 #define FOURCC_schm     GST_MAKE_FOURCC('s','c','h','m')
 #define FOURCC_schi     GST_MAKE_FOURCC('s','c','h','i')
+#define FOURCC_senc     GST_MAKE_FOURCC('s','e','n','c')
 
 /* Common Encryption */
 #define FOURCC_pssh     GST_MAKE_FOURCC('p','s','s','h')
diff --git a/gst/isomp4/qtdemux_types.c b/gst/isomp4/qtdemux_types.c
index 5c301b9..c41fc89 100755
--- a/gst/isomp4/qtdemux_types.c
+++ b/gst/isomp4/qtdemux_types.c
@@ -233,6 +233,7 @@ static const QtNodeType qt_node_types[] = {
   {FOURCC_dvcC, "DOVI in dvcC box", 0},
   {FOURCC_dvvC, "DOVI in dvvC box", 0},
   {FOURCC_dvwC, "DOVI in dvwC box", 0},
+  {FOURCC_senc, "sample encryption box", 0},
   {0, "unknown", 0,},
 };
 
