Index: gst-plugins-good-1.16.3/gst/matroska/matroska-demux.c
===================================================================
--- gst-plugins-good-1.16.3.orig/gst/matroska/matroska-demux.c
+++ gst-plugins-good-1.16.3/gst/matroska/matroska-demux.c
@@ -287,7 +287,7 @@ gst_matroska_demux_init (GstMatroskaDemu
   GST_OBJECT_FLAG_SET (demux, GST_ELEMENT_FLAG_INDEXABLE);
 
   demux->flowcombiner = gst_flow_combiner_new ();
-
+  g_queue_init(&demux->pending_protection_event_queue);
   /* finish off */
   gst_matroska_demux_reset (GST_ELEMENT (demux));
 }
@@ -362,6 +362,9 @@ gst_matroska_demux_reset (GstElement * e
   demux->deferred_seek_event = NULL;
   demux->deferred_seek_pad = NULL;
 
+  g_queue_foreach (&demux->pending_protection_event_queue, (GFunc) gst_event_unref,
+    NULL);
+  g_queue_clear (&demux->pending_protection_event_queue);
   gst_flow_combiner_clear (demux->flowcombiner);
 }
 
@@ -754,6 +757,15 @@ beach:
   return ret;
 }
 
+void 
+gst_matroska_demux_copy_pending_protection_event(gpointer       data,
+                                                 gpointer       user_data)
+{
+  GstEvent *event = (GstEvent *)data;
+  GstMatroskaTrackContext *context = (GstMatroskaTrackContext *)user_data;
+  g_queue_push_tail (&context->protection_event_queue, gst_event_ref(event));
+}
+
 static GstFlowReturn
 gst_matroska_demux_parse_stream (GstMatroskaDemux * demux, GstEbmlRead * ebml,
     GstMatroskaTrackContext ** dest_context)
@@ -797,7 +809,7 @@ gst_matroska_demux_parse_stream (GstMatr
   context->tags = gst_tag_list_new_empty ();
   g_queue_init (&context->protection_event_queue);
   context->protection_info = NULL;
-
+  g_queue_foreach(&demux->pending_protection_event_queue, gst_matroska_demux_copy_pending_protection_event, context);
   GST_DEBUG_OBJECT (demux, "Parsing a TrackEntry (%d tracks parsed so far)",
       demux->common.num_streams);
 
@@ -5965,6 +5977,11 @@ gst_matroska_demux_handle_sink_event (Gs
       "have event type %s: %p on sink pad", GST_EVENT_TYPE_NAME (event), event);
 
   switch (GST_EVENT_TYPE (event)) {
+    case GST_EVENT_PROTECTION:
+    {
+      g_queue_push_tail (&demux->pending_protection_event_queue, event);
+      break;
+    }
     case GST_EVENT_SEGMENT:
     {
       const GstSegment *segment;
Index: gst-plugins-good-1.16.3/gst/matroska/matroska-demux.h
===================================================================
--- gst-plugins-good-1.16.3.orig/gst/matroska/matroska-demux.h
+++ gst-plugins-good-1.16.3/gst/matroska/matroska-demux.h
@@ -128,6 +128,10 @@ typedef struct _GstMatroskaDemux {
 
   /* Cached upstream length (default G_MAXUINT64) */
   guint64	           cached_length;
+
+  /* Pending protection event need pass to downstream */
+  GQueue                   pending_protection_event_queue;
+
 } GstMatroskaDemux;
 
 typedef struct _GstMatroskaDemuxClass {
Index: gst-plugins-good-1.16.3/gst/matroska/matroska-read-common.c
===================================================================
--- gst-plugins-good-1.16.3.orig/gst/matroska/matroska-read-common.c
+++ gst-plugins-good-1.16.3/gst/matroska/matroska-read-common.c
@@ -448,7 +448,7 @@ gst_matroska_parse_protection_meta (gpoi
 
   /* Unencrypted buffer */
   if (!(signal_byte & GST_MATROSKA_BLOCK_ENCRYPTED)) {
-    return TRUE;
+    goto beach;
   }
 
   /* Encrypted buffer */
@@ -591,7 +591,7 @@ gst_matroska_parse_protection_meta (gpoi
   } else {
     gst_structure_set (info_protect, "subsample_count", G_TYPE_UINT, 0, NULL);
   }
-
+beach:
   gst_byte_reader_get_data (&reader, 0, (const guint8 **) data_out);
   *size_out = gst_byte_reader_get_remaining (&reader);
   return TRUE;
