diff --git a/gst/isomp4/qtdemux.c b/gst/isomp4/qtdemux.c
index 382b9e8..2dadc0c 100755
--- a/gst/isomp4/qtdemux.c
+++ b/gst/isomp4/qtdemux.c
@@ -6358,7 +6358,8 @@ gst_qtdemux_loop_state_movie (GstQTDemux * qtdemux)
     position = stream->time_position;
 
     /* position of -1 is EOS */
-    if (position != GST_CLOCK_TIME_NONE && position < min_time) {
+    if ((position != GST_CLOCK_TIME_NONE && position < min_time)
+     || (stream->discont)) {
       min_time = position;
       target_stream = stream;
     }
