diff --git a/gst/isomp4/qtdemux.c b/gst/isomp4/qtdemux.c
index c234f5e..c24b4f6 100644
--- a/gst/isomp4/qtdemux.c
+++ b/gst/isomp4/qtdemux.c
@@ -5325,14 +5325,14 @@ gst_qtdemux_advance_sample (GstQTDemux * qtdemux, QtDemuxStream * stream)
       GST_TIME_ARGS (segment->media_stop));
 
   /* see if we are past the segment */
-  if (G_UNLIKELY (QTSAMPLE_DTS (stream, sample) >= segment->media_stop))
+  if (G_UNLIKELY (QTSAMPLE_PTS (stream, sample) >= segment->media_stop))
     goto next_segment;
 
-  if (QTSAMPLE_DTS (stream, sample) >= segment->media_start) {
+  if (QTSAMPLE_PTS (stream, sample) >= segment->media_start) {
     /* inside the segment, update time_position, looks very familiar to
      * GStreamer segments, doesn't it? */
     stream->time_position =
-        QTSAMPLE_DTS (stream, sample) - segment->media_start + segment->time;
+        QTSAMPLE_PTS (stream, sample) - segment->media_start + segment->time;
   } else {
     /* not yet in segment, time does not yet increment. This means
      * that we are still prerolling keyframes to the decoder so it can
