diff --git a/sys/v4l2/gstv4l2bufferpool.c b/sys/v4l2/gstv4l2bufferpool.c
index 1586608..ea1e32c 100755
--- a/sys/v4l2/gstv4l2bufferpool.c
+++ b/sys/v4l2/gstv4l2bufferpool.c
@@ -1013,20 +1013,20 @@ gst_v4l2_buffer_pool_stop (GstBufferPool * bpool)
     pool->group_released_handler = 0;
   }
 
-  if (pool->other_pool) {
-    gst_buffer_pool_set_active (pool->other_pool, FALSE);
-    gst_object_unref (pool->other_pool);
-    pool->other_pool = NULL;
-  }
-
-  if (!pool->orphaned)
-    gst_v4l2_buffer_pool_streamoff (pool);
+  gst_v4l2_buffer_pool_streamoff (pool);
 
   ret = GST_BUFFER_POOL_CLASS (parent_class)->stop (bpool);
 
   if (ret)
     ret = gst_v4l2_buffer_pool_vallocator_stop (pool);
 
+  GST_DEBUG_OBJECT (pool, "stopping other_pool");
+  if (pool->other_pool) {
+    gst_buffer_pool_set_active (pool->other_pool, FALSE);
+    gst_object_unref (pool->other_pool);
+    pool->other_pool = NULL;
+  }
+
   return ret;
 }
 
diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index 7bb9534..a2aef8e 100755
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -573,6 +573,8 @@ gst_v4l2_object_destroy (GstV4l2Object * v4l2object)
     gst_structure_free (v4l2object->extra_controls);
   }
 
+  gst_poll_free (v4l2object->poll);
+
   g_free (v4l2object);
 }
 
@@ -4652,8 +4654,6 @@ gst_v4l2_object_stop (GstV4l2Object * v4l2object)
     gst_object_unref (pool);
   }
 
-  gst_poll_free (pool);
-
   GST_V4L2_SET_INACTIVE (v4l2object);
 
 done:
