Index: bluez-5.55/src/adapter.c
===================================================================
--- bluez-5.55.orig/src/adapter.c
+++ bluez-5.55/src/adapter.c
@@ -7830,6 +7830,14 @@ static void dev_disconnected(struct btd_
 	if (device) {
 		adapter_remove_connection(adapter, device, addr->type);
 		disconnect_notify(device, reason);
+		DBG("device found, checking reason and addr type");
+		if (reason == 4 //mgmt auth failure
+			&& addr->type != BDADDR_BREDR
+			&& device_is_paired(device, addr->type)) {
+			DBG("BLE device and auth failure, remove deivce");
+			btd_adapter_remove_device(adapter, device);
+		}
+
 	}
 
 	bonding_attempt_complete(adapter, &addr->bdaddr, addr->type,
