Index: core/adb/adb.c
===================================================================
--- core.orig/adb/adb.c
+++ core/adb/adb.c
@@ -1414,7 +1414,7 @@ int adb_main(int is_daemon, int server_p
         printf("using port=%d\n", port);
         // listen on TCP port specified by service.adb.tcp.port property
         local_init(port);
-    } else if (!usb) {
+    } else { //if (!usb) {
         // listen on default port
         local_init(DEFAULT_ADB_LOCAL_TRANSPORT_PORT);
     }
