diff --git a/gst/mpegtsdemux/gstmpegdesc.h b/gst/mpegtsdemux/gstmpegdesc.h
index 65fb44f..99d597b 100644
--- a/gst/mpegtsdemux/gstmpegdesc.h
+++ b/gst/mpegtsdemux/gstmpegdesc.h
@@ -224,6 +224,7 @@
 #define DRF_ID_DTS1       0x44545331
 #define DRF_ID_DTS2       0x44545332
 #define DRF_ID_DTS3       0x44545333
+#define DRF_ID_DTS4       0x44545348
 #define DRF_ID_S302M      0x42535344
 #define DRF_ID_TSHV       0x54534856
 #define DRF_ID_AC3        0x41432d33
diff --git a/gst/mpegtsdemux/tsdemux.c b/gst/mpegtsdemux/tsdemux.c
index e125381..68cac44 100644
--- a/gst/mpegtsdemux/tsdemux.c
+++ b/gst/mpegtsdemux/tsdemux.c
@@ -1495,6 +1495,7 @@ create_pad_for_stream (MpegTSBase * base, MpegTSBaseStream * bstream,
         case DRF_ID_DTS1:
         case DRF_ID_DTS2:
         case DRF_ID_DTS3:
+        case DRF_ID_DTS4:
           /* SMPTE registered DTS */
           is_private = TRUE;
           caps = gst_caps_new_empty_simple ("audio/x-dts");
