diff --git a/ext/wayland/wlwindow.c b/ext/wayland/wlwindow.c
old mode 100644
new mode 100755
index c3dc61d..49626c4
--- a/ext/wayland/wlwindow.c
+++ b/ext/wayland/wlwindow.c
@@ -319,10 +319,11 @@ gst_wl_window_new_toplevel (GstWlDisplay * display, const GstVideoInfo * info,
     goto error;
   }
 
-  /* set the initial size to be the same as the reported video size */
-  width =
-      gst_util_uint64_scale_int_round (info->width, info->par_n, info->par_d);
-  gst_wl_window_set_render_rectangle (window, 0, 0, width, info->height);
+  if (!(display->xdg_wm_base && fullscreen)) {
+    /* set the initial size to be the same as the reported video size */
+    width = gst_util_uint64_scale_int_round (info->width, info->par_n, info->par_d);
+    gst_wl_window_set_render_rectangle (window, 0, 0, width, info->height);
+  }
 
   return window;
 
@@ -550,6 +551,9 @@ gst_wl_window_set_render_rectangle (GstWlWindow * window, gint x, gint y,
 
   gst_wl_window_update_borders (window);
 
+  if (!window->configured)
+    return;
+
   if (window->video_width != 0) {
     wl_subsurface_set_sync (window->video_subsurface);
     gst_wl_window_resize_video_surface (window, TRUE);
