diff --git a/gst/videoparsers/gsth264parse.c b/gst/videoparsers/gsth264parse.c
index 11581b0..dde78dc 100644
--- a/gst/videoparsers/gsth264parse.c
+++ b/gst/videoparsers/gsth264parse.c
@@ -1379,6 +1379,13 @@ gst_h264_parse_handle_frame (GstBaseParse * parse,
     goto end;
   }
 
+  if (current_off >= size)
+  {
+    GST_DEBUG_OBJECT (h264parse, "Need more data, current_off %d, size %d",
+    current_off, size);
+    goto more;
+  }
+
   g_assert (current_off < size);
   GST_DEBUG_OBJECT (h264parse, "last parse position %d", current_off);
 
