From 2f8b26acd2814cd707d11e7f55b0e0eaa0cad65d Mon Sep 17 00:00:00 2001
From: "kaiqiang.xiang" <kaiqiang.xiang@amlogic.com>
Date: Tue, 17 May 2022 20:17:31 +0800
Subject: [PATCH] modify bug fragment loss

---
 ext/hls/gsthlsdemux.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/ext/hls/gsthlsdemux.c b/ext/hls/gsthlsdemux.c
index 7dc0a93..d3ea6ee 100644
--- a/ext/hls/gsthlsdemux.c
+++ b/ext/hls/gsthlsdemux.c
@@ -1062,7 +1062,8 @@ gst_hls_demux_advance_fragment (GstAdaptiveDemuxStream * stream)
 
   m3u8 = gst_hls_demux_stream_get_m3u8 (hlsdemux_stream);
 
-  gst_m3u8_advance_fragment (m3u8, stream->demux->segment.rate > 0);
+  if(!GST_ADAPTIVE_DEMUX_STREAM_NEED_HEADER (stream))
+    gst_m3u8_advance_fragment (m3u8, stream->demux->segment.rate > 0);
   hlsdemux_stream->reset_pts = FALSE;
 
   return GST_FLOW_OK;
-- 
2.25.1

