diff --git a/gst/mpegtsdemux/gstmpegdesc.h b/gst/mpegtsdemux/gstmpegdesc.h
index 99d597b..36f0dba 100644
--- a/gst/mpegtsdemux/gstmpegdesc.h
+++ b/gst/mpegtsdemux/gstmpegdesc.h
@@ -236,5 +236,6 @@
 #define DRF_ID_OPUS       0x4f707573
 #define DRF_ID_EAC3       0x45414333   /* defined in A/52 Annex G */
 #define DRF_ID_AC4        0x41432D34   /* defined in ETSI TS 103 190-2 Annex D */
+#define DRF_ID_DolbyVision  0x444f5649
 
 #endif /* __GST_MPEG_DESC_H__ */
diff --git a/gst/mpegtsdemux/tsdemux.c b/gst/mpegtsdemux/tsdemux.c
index 3532592..cf05983 100644
--- a/gst/mpegtsdemux/tsdemux.c
+++ b/gst/mpegtsdemux/tsdemux.c
@@ -1683,6 +1683,7 @@ create_pad_for_stream (MpegTSBase * base, MpegTSBaseStream * bstream,
           }
           break;
         case DRF_ID_HEVC:
+        case DRF_ID_DolbyVision:
           is_video = TRUE;
           caps = gst_caps_new_simple ("video/x-h265",
               "stream-format", G_TYPE_STRING, "byte-stream", NULL);
