diff --git a/gst/mpegtsdemux/mpegtspacketizer.c b/gst/mpegtsdemux/mpegtspacketizer.c
index 2d6e404..ef235f8 100644
--- a/gst/mpegtsdemux/mpegtspacketizer.c
+++ b/gst/mpegtsdemux/mpegtspacketizer.c
@@ -1898,8 +1898,9 @@ _set_current_group (MpegTSPCR * pcrtable,
     } else if (lastpcr < pcr - 500 * PCR_MSECOND) {
       GST_WARNING ("GAP detected. diff %" GST_TIME_FORMAT,
           GST_TIME_ARGS (PCRTIME_TO_GSTTIME (pcr - lastpcr)));
-      /* The previous group closed at the raw last_pcr diff (+500ms for safety) */
-      pcr_offset += prev->values[prev->last_value].pcr + 500 * PCR_MSECOND;
+     /*pcr_offset need to add gap time of pcr to lastpcr, because prev->values[prev->last_value].pcr is
+       always 0 that cause calcute current pts issue*/
+     pcr_offset += pcr - lastpcr;
     } else
       /* Normal continuation (contiguous in time) */
       pcr_offset += pcr - prev->first_pcr;
