diff --git a/gst/videoparsers/gsth264parse.c b/gst/videoparsers/gsth264parse.c
index dde78dc..364b614 100644
--- a/gst/videoparsers/gsth264parse.c
+++ b/gst/videoparsers/gsth264parse.c
@@ -2605,14 +2605,14 @@ gst_h264_parse_get_timestamp (GstH264Parse * h264parse,
           (h264parse->sei_cpb_removal_delay * GST_SECOND,
           sps->vui_parameters.num_units_in_tick,
           sps->vui_parameters.time_scale);
-    } else {
+    } //else {
       /* If no upstream timestamp is given, we write in new timestamp */
-      upstream = h264parse->dts = h264parse->ts_trn_nb +
-          (GstClockTime) gst_util_uint64_scale
-          (h264parse->sei_cpb_removal_delay * GST_SECOND,
-          sps->vui_parameters.num_units_in_tick,
-          sps->vui_parameters.time_scale);
-    }
+      //upstream = h264parse->dts = h264parse->ts_trn_nb +
+      //    (GstClockTime) gst_util_uint64_scale
+      //    (h264parse->sei_cpb_removal_delay * GST_SECOND,
+      //    sps->vui_parameters.num_units_in_tick,
+      //    sps->vui_parameters.time_scale);
+      //}
   } else {
     GstClockTime dur;
 
