diff --git a/gst/mpegtsdemux/tsdemux.c b/gst/mpegtsdemux/tsdemux.c
index 2b4947f..634d466 100644
--- a/gst/mpegtsdemux/tsdemux.c
+++ b/gst/mpegtsdemux/tsdemux.c
@@ -2017,6 +2017,14 @@ done:
     gst_pad_set_caps (pad, caps);
     gst_pad_set_query_function (pad, gst_ts_demux_srcpad_query);
     gst_pad_set_event_function (pad, gst_ts_demux_srcpad_event);
+    if (is_video) {
+        GstStructure *s;
+        GstEvent *event;
+        s = gst_structure_new_empty ("AML-SET-MAX-BYTE-SIZE");
+        event = gst_event_new_custom (GST_EVENT_CUSTOM_DOWNSTREAM_STICKY, s);
+        GST_DEBUG("Send SET-MAX-BYTE-SIZE Event");
+        gst_pad_push_event (pad, event);
+    }
   }
 
   g_free (name);
