diff --git a/gst/mpegtsdemux/tsdemux.c b/gst/mpegtsdemux/tsdemux.c
index d510eec..8582b2f 100644
--- a/gst/mpegtsdemux/tsdemux.c
+++ b/gst/mpegtsdemux/tsdemux.c
@@ -3619,6 +3619,7 @@ gst_ts_demux_push_pending_data (GstTSDemux * demux, TSDemuxStream * stream,
         /*GST_DEBUG("attaching secure IV to output buffer");*/
         gst_buffer_add_protection_meta(buffer, drm_info_struct);
     }
+    gst_buffer_unref(iv_buffer);
   }
 
   if (buffer) {
