diff --git a/gst/mpegtsdemux/mpegtspacketizer.c b/gst/mpegtsdemux/mpegtspacketizer.c
index ef235f8..9960a29 100644
--- a/gst/mpegtsdemux/mpegtspacketizer.c
+++ b/gst/mpegtsdemux/mpegtspacketizer.c
@@ -2375,8 +2375,11 @@ mpegts_packetizer_pts_to_ts (MpegTSPacketizer2 * packetizer,
           pts - PCRTIME_TO_GSTTIME (refpcr) + PCRTIME_TO_GSTTIME (refpcroffset);
     else
       GST_WARNING ("No groups, can't calculate timestamp");
-  } else
-    GST_WARNING ("Not enough information to calculate proper timestamp");
+  } else {
+    //AML MOD FLOW
+    GST_WARNING ("Not enough information to calculate proper timestamp use origin pts %" GST_TIME_FORMAT, GST_TIME_ARGS (pts));
+    res = pts;
+  }
 
   PACKETIZER_GROUP_UNLOCK (packetizer);
 
