diff --git a/gst/mpegtsdemux/tsdemux.c b/gst/mpegtsdemux/tsdemux.c
index 8582b2f..5f06a65 100644
--- a/gst/mpegtsdemux/tsdemux.c
+++ b/gst/mpegtsdemux/tsdemux.c
@@ -521,6 +521,7 @@ gst_ts_demux_init (GstTSDemux * demux)
   demux->flowcombiner = gst_flow_combiner_new ();
   demux->requested_program_number = -1;
   demux->program_number = -1;
+  demux->no_check_pcr_count = 0;
   demux->latency = DEFAULT_LATENCY;
   gst_ts_demux_reset (base);
 
@@ -3496,8 +3497,7 @@ gst_ts_demux_push_pending_data (GstTSDemux * demux, TSDemuxStream * stream,
     if (G_UNLIKELY (stream->pending_ts && !check_pending_buffers (demux))) {
       //according to prop process issue ts file, in advance goto error
       if (!demux->no_check_pcr_illegal) {
-        static int count = 0;
-        if (count++ > 200) {
+        if (demux->no_check_pcr_count++ > 200) {
           res = GST_FLOW_ERROR;
           GST_DEBUG("process pcr error, return gst_flow_error");
           goto beach;
diff --git a/gst/mpegtsdemux/tsdemux.h b/gst/mpegtsdemux/tsdemux.h
index e5b5666..dc4747a 100644
--- a/gst/mpegtsdemux/tsdemux.h
+++ b/gst/mpegtsdemux/tsdemux.h
@@ -77,6 +77,7 @@ struct _GstTSDemux
   guint program_number;
   gboolean emit_statistics;
   gboolean no_check_pcr_illegal;
+  guint no_check_pcr_count;
   gboolean send_scte35_events;
   gint latency; /* latency in ms */
 
