diff --git a/gst/mpegtsdemux/mpegtsbase.c b/gst/mpegtsdemux/mpegtsbase.c
index 85d5afd..10628af 100644
--- a/gst/mpegtsdemux/mpegtsbase.c
+++ b/gst/mpegtsdemux/mpegtsbase.c
@@ -1903,6 +1903,10 @@ mpegts_base_handle_seek_event (MpegTSBase * base, GstPad * pad,
     goto done;
   }
 
+  if (flags & GST_SEEK_FLAG_TRICKMODE_NO_AUDIO)
+    base->no_audio_set = TRUE;
+  else
+    base->no_audio_set = FALSE;
 
   /* If the subclass can seek, do that */
   ret = klass->seek (base, event);
diff --git a/gst/mpegtsdemux/mpegtsbase.h b/gst/mpegtsdemux/mpegtsbase.h
index 3d463f4..fab6d6a 100644
--- a/gst/mpegtsdemux/mpegtsbase.h
+++ b/gst/mpegtsdemux/mpegtsbase.h
@@ -170,6 +170,9 @@ struct _MpegTSBase {
    * add/remove streams at any point in time */
   gboolean streams_aware;
 
+  /* save seek flag */
+  gboolean no_audio_set;
+
   /* Do not use the PCR stream for timestamp calculation. Useful for
    * streams with broken/invalid PCR streams. */
   gboolean ignore_pcr;
diff --git a/gst/mpegtsdemux/tsdemux.c b/gst/mpegtsdemux/tsdemux.c
index 5f06a65..ea36de7 100644
--- a/gst/mpegtsdemux/tsdemux.c
+++ b/gst/mpegtsdemux/tsdemux.c
@@ -2901,6 +2901,7 @@ calculate_and_push_newsegment (GstTSDemux * demux, TSDemuxStream * stream,
       GstClockTime base =
           seg->base + seg->position - (seg->start + seg->offset);
       GstClockTime stop = seg->stop;
+      MpegTSBase *tsbase = (MpegTSBase *) demux;
 
       gst_segment_init (seg, GST_FORMAT_TIME);
       seg->start = firstts;
@@ -2909,6 +2910,8 @@ calculate_and_push_newsegment (GstTSDemux * demux, TSDemuxStream * stream,
       seg->time = firstts;
       seg->rate = demux->rate;
       seg->base = base;
+      if (tsbase->no_audio_set)
+        seg->flags |= GST_SEGMENT_FLAG_TRICKMODE_NO_AUDIO;
     }
   } else if (base->out_segment.start < firstts) {
     /* Take into account the offset to the first buffer timestamp */
