diff --git a/gst-libs/gst/adaptivedemux/gstadaptivedemux.c b/gst-libs/gst/adaptivedemux/gstadaptivedemux.c
index 08d8d81..404f959 100644
--- a/gst-libs/gst/adaptivedemux/gstadaptivedemux.c
+++ b/gst-libs/gst/adaptivedemux/gstadaptivedemux.c
@@ -1912,7 +1912,7 @@ gst_adaptive_demux_src_event (GstPad * pad, GstObject * parent,
       GST_DEBUG_OBJECT (demux, "unlock  demux lock after set isSeeking to TRUE");
       g_print ("unlock  demux lock after set isSeeking to TRUE\n");
       GST_MANIFEST_UNLOCK (demux);
-      gint64 end_time;
+      gint64 end_time, out_time = 0;
       while(demux->priv->isSwitching)
       {
         g_mutex_lock (&demux->priv->adaptive_switch_lock);
@@ -1927,6 +1927,8 @@ gst_adaptive_demux_src_event (GstPad * pad, GstObject * parent,
         }
         GST_DEBUG_OBJECT (demux, "continue waiting cond : adaptive_switch_cond or isSwitching == FALSE");
         g_mutex_unlock (&demux->priv->adaptive_switch_lock);
+        if(100 == out_time)  break;
+        out_time++;
       }
 
       gboolean ret = gst_adaptive_demux_handle_seek_event (demux, pad, event);
