diff --git a/gst/videoparsers/gstmpegvideoparse.c b/gst/videoparsers/gstmpegvideoparse.c
index f8ef31a..7f64934 100644
--- a/gst/videoparsers/gstmpegvideoparse.c
+++ b/gst/videoparsers/gstmpegvideoparse.c
@@ -576,9 +576,16 @@ gst_mpegv_parse_process_sc (GstMpegvParse * mpvparse,
     header.offset = mpvparse->pic_offset;
     header.size = info->size - mpvparse->pic_offset;
     if (gst_mpeg_video_packet_parse_picture_header (&header, &mpvparse->pichdr))
+    {
       GST_LOG_OBJECT (mpvparse, "picture_coding_type %d (%s), ending"
           "frame of size %d", mpvparse->pichdr.pic_type,
           picture_type_name (mpvparse->pichdr.pic_type), off - 4);
+      if (!(mpvparse->got_first_I_frame) && (1 == mpvparse->pichdr.pic_type))
+      {
+        GST_DEBUG_OBJECT (mpvparse, "got the firset I frame, set mpvparse->got_first_I_frame to TRUE");
+        mpvparse->got_first_I_frame = TRUE;
+      }
+    }
     else
       GST_LOG_OBJECT (mpvparse, "Couldn't parse picture at offset %d",
           mpvparse->pic_offset);
@@ -637,6 +644,12 @@ gst_mpegv_parse_handle_frame (GstBaseParse * parse,
   gboolean need_more = FALSE;
   GstMapInfo map;
 
+  if (GST_BUFFER_FLAG_IS_SET (buf, GST_BUFFER_FLAG_DISCONT) && mpvparse->got_first_I_frame == TRUE)
+  {
+    GST_DEBUG_OBJECT (mpvparse, "BUFFER_FLAG_DISCONT, set got_first_I_frame to FALSE");
+    mpvparse->got_first_I_frame = FALSE;
+  }
+
   update_frame_parsing_status (mpvparse, frame);
 
   gst_buffer_map (buf, &map, GST_MAP_READ);
@@ -974,6 +987,12 @@ gst_mpegv_parse_pre_push_frame (GstBaseParse * parse, GstBaseParseFrame * frame)
   GstMpegVideoQuantMatrixExt *quant_ext = NULL;
   GstBuffer *parse_buffer = NULL;
 
+  if (!(mpvparse->got_first_I_frame))
+  {
+    GST_DEBUG_OBJECT (mpvparse, "Drop non-I frame");
+    return GST_BASE_PARSE_FLOW_DROPPED;
+  }
+
   /* tag sending done late enough in hook to ensure pending events
    * have already been sent */
 
diff --git a/gst/videoparsers/gstmpegvideoparse.h b/gst/videoparsers/gstmpegvideoparse.h
index 4052beb..b718fc9 100644
--- a/gst/videoparsers/gstmpegvideoparse.h
+++ b/gst/videoparsers/gstmpegvideoparse.h
@@ -86,6 +86,7 @@ struct _GstMpegvParse {
   gboolean seqdispext_updated;
   gboolean picext_updated;
   gboolean quantmatrext_updated;
+  gboolean got_first_I_frame;
 
   GstVideoParseUserData user_data;
 
