diff --git a/ext/hls/gsthlsdemux.c b/ext/hls/gsthlsdemux.c
index 8cf6c2a..37e2b53 100644
--- a/ext/hls/gsthlsdemux.c
+++ b/ext/hls/gsthlsdemux.c
@@ -1563,7 +1563,7 @@ retry:
   uri = gst_m3u8_get_uri (demux->current_variant->m3u8);
   main_uri = gst_adaptive_demux_get_manifest_ref_uri (adaptive_demux);
   download =
-      gst_uri_downloader_fetch_uri (adaptive_demux->downloader, uri, main_uri,
+      gst_uri_downloader_fetch_uri (adaptive_demux->downloader, uri, NULL,
       TRUE, TRUE, TRUE, err);
   if (download == NULL) {
     gchar *base_uri;
