diff --git a/ext/hls/gsthlsdemux.c b/ext/hls/gsthlsdemux.c
index 37e2b53..7738922 100644
--- a/ext/hls/gsthlsdemux.c
+++ b/ext/hls/gsthlsdemux.c
@@ -1032,6 +1032,7 @@ gst_hls_demux_handle_buffer (GstAdaptiveDemux * demux,
     hls_stream->pending_pcr_buffer = NULL;
   }
 
+#if 0 //delete for airplay youtube play slowly
   if (!gst_hlsdemux_tsreader_find_pcrs (&hls_stream->tsreader, &buffer,
           &first_pcr, &last_pcr, &tags)
       && !at_eos) {
@@ -1039,6 +1040,7 @@ gst_hls_demux_handle_buffer (GstAdaptiveDemux * demux,
     hls_stream->pending_pcr_buffer = buffer;
     return GST_FLOW_OK;
   }
+#endif
 
   if (tags) {
     gst_adaptive_demux_stream_set_tags (stream, tags);
