diff --git a/gst/mpegtsdemux/tsdemux.c b/gst/mpegtsdemux/tsdemux.c
index 35317e3..39e36fb 100644
--- a/gst/mpegtsdemux/tsdemux.c
+++ b/gst/mpegtsdemux/tsdemux.c
@@ -3614,6 +3614,15 @@ gst_ts_demux_push_pending_data (GstTSDemux * demux, TSDemuxStream * stream,
       GST_TIME_FORMAT, (buffer_list ? "list" : ""), GST_TIME_ARGS (stream->pts),
       GST_TIME_ARGS (stream->dts));
 
+  GstClockTime dur;
+  gst_ts_demux_get_duration (demux, &dur);
+  if( dur<stream->pts )
+  {
+    GST_ELEMENT_ERROR(demux, RESOURCE, FAILED, ("GStreamer encountered a general resource error"), ("GStreamer encountered a general resource error"));
+    res = GST_FLOW_ERROR;
+    goto beach;
+  }
+
   if (GST_CLOCK_TIME_IS_VALID (stream->dts)) {
     if (stream->dts > base->out_segment.position)
       base->out_segment.position = stream->dts;
