diff --git a/gst/mpegtsdemux/tsdemux.c b/gst/mpegtsdemux/tsdemux.c
index 7a5fc78..4569505 100644
--- a/gst/mpegtsdemux/tsdemux.c
+++ b/gst/mpegtsdemux/tsdemux.c
@@ -3629,7 +3629,7 @@ gst_ts_demux_push_pending_data (GstTSDemux * demux, TSDemuxStream * stream,
   {
       GstClockTime dur;
       gst_ts_demux_get_duration (demux, &dur);
-      if( dur < stream->pts )
+      if( (dur+3*GST_SECOND < stream->pts) && (-1 != stream->pts) )
       {
          GST_ELEMENT_ERROR(demux, RESOURCE, FAILED, ("Duration is shorter than the start time resource error"),
              ("Duration is shorter than the start time resource error"));
