diff --git a/ext/hls/gsthlsdemux.c b/ext/hls/gsthlsdemux.c
index 7738922..d2ba84a 100644
--- a/ext/hls/gsthlsdemux.c
+++ b/ext/hls/gsthlsdemux.c
@@ -967,7 +967,7 @@ gst_hls_demux_handle_buffer (GstAdaptiveDemux * demux,
   GstHLSDemuxStream *hls_stream = GST_HLS_DEMUX_STREAM_CAST (stream);   // FIXME: pass HlsStream into function
   GstHLSDemux *hlsdemux = GST_HLS_DEMUX_CAST (demux);
   GstClockTime first_pcr, last_pcr;
-  GstTagList *tags;
+  GstTagList *tags = NULL;
 
   if (buffer == NULL)
     return GST_FLOW_OK;
