diff --git a/ext/hls/gsthlsdemux.c b/ext/hls/gsthlsdemux.c
index 05e0205..528e45c 100644
--- a/ext/hls/gsthlsdemux.c
+++ b/ext/hls/gsthlsdemux.c
@@ -867,7 +867,7 @@ gst_hls_demux_get_key (GstHLSDemux * demux, const gchar * key_url,
 
   key_fragment =
       gst_uri_downloader_fetch_uri (GST_ADAPTIVE_DEMUX (demux)->downloader,
-      key_url, referer, FALSE, FALSE, allow_cache, &err);
+      key_url, NULL, FALSE, FALSE, allow_cache, &err);
 
   if (key_fragment == NULL) {
     GST_WARNING_OBJECT (demux, "Failed to download key to decrypt data: %s",
