diff --git a/gst/mpegdemux/gstmpegdemux.c b/gst/mpegdemux/gstmpegdemux.c
index b5fa342..8a6ebca 100644
--- a/gst/mpegdemux/gstmpegdemux.c
+++ b/gst/mpegdemux/gstmpegdemux.c
@@ -665,13 +665,15 @@ gst_ps_demux_send_segment (GstPsDemux * demux, GstPsStream * stream,
     /* we should be in sync with downstream, so start from our segment notion,
      * which also includes proper base_time etc, tweak it a bit and send */
     gst_segment_copy_into (&demux->src_segment, &segment);
+
+    /*
     if (GST_CLOCK_TIME_IS_VALID (demux->base_time)) {
       if (GST_CLOCK_TIME_IS_VALID (segment.start))
         segment.start += demux->base_time;
       if (GST_CLOCK_TIME_IS_VALID (segment.stop))
         segment.stop += demux->base_time;
       segment.time = segment.start - demux->base_time;
-    }
+    }*/
 
     segment_event = gst_event_new_segment (&segment);
     if (demux->segment_seqnum)
@@ -708,9 +710,19 @@ gst_ps_demux_send_data (GstPsDemux * demux, GstPsStream * stream,
 
   /* timestamps */
   if (G_UNLIKELY (demux->next_pts != G_MAXUINT64))
-    pts = MPEGTIME_TO_GSTTIME (demux->next_pts);
+  {
+    if ( demux->next_pts > 0 && demux->first_scr > 0 && demux->next_pts > demux->first_scr)
+      pts = MPEGTIME_TO_GSTTIME (demux->next_pts - demux->first_scr);
+    else
+      pts = MPEGTIME_TO_GSTTIME (demux->next_pts);
+  }
   if (G_UNLIKELY (demux->next_dts != G_MAXUINT64))
-    dts = MPEGTIME_TO_GSTTIME (demux->next_dts);
+  {
+    if ( demux->next_dts > 0 && demux->first_scr > 0 && demux->next_dts > demux->first_scr)
+      dts = MPEGTIME_TO_GSTTIME (demux->next_dts - demux->first_scr);
+    else
+      dts = MPEGTIME_TO_GSTTIME (demux->next_dts);
+  }
 
   gst_ps_demux_send_segment (demux, stream, pts);
 
