diff --git a/gst/mpegdemux/gstmpegdemux.c b/gst/mpegdemux/gstmpegdemux.c
index 8a6ebca..805c3ec 100644
--- a/gst/mpegdemux/gstmpegdemux.c
+++ b/gst/mpegdemux/gstmpegdemux.c
@@ -2954,8 +2954,11 @@ gst_ps_sink_get_duration (GstPsDemux * demux)
           demux->last_pts != G_MAXUINT64)) {
     /* update the src segment */
     demux->src_segment.format = GST_FORMAT_TIME;
-    demux->src_segment.start =
+    if (MPEGTIME_TO_GSTTIME (demux->first_pts) > demux->base_time)
+      demux->src_segment.start =
         MPEGTIME_TO_GSTTIME (demux->first_pts) - demux->base_time;
+    else
+      demux->src_segment.start = MPEGTIME_TO_GSTTIME (demux->first_pts);
     demux->src_segment.stop = -1;
     gst_segment_set_duration (&demux->src_segment, GST_FORMAT_TIME,
         MPEGTIME_TO_GSTTIME (demux->last_pts - demux->first_pts));
