diff --git a/ext/hls/gsthlsdemux.c b/ext/hls/gsthlsdemux.c
index 201cc72..c8499ec 100644
--- a/ext/hls/gsthlsdemux.c
+++ b/ext/hls/gsthlsdemux.c
@@ -1279,6 +1279,13 @@ gst_hls_demux_update_fragment_info (GstAdaptiveDemuxStream * stream)
     stream->fragment.timestamp = GST_CLOCK_TIME_NONE;
   }
 
+  //AML patch: should for live and ts
+  if (gst_hls_variant_stream_is_live (hlsdemux->current_variant))
+  {
+    stream->fragment.timestamp = sequence_pos;
+  }
+  //AML patch end
+
   g_free (hlsdemux_stream->current_key);
   hlsdemux_stream->current_key = g_strdup (file->key);
   g_free (hlsdemux_stream->current_iv);
diff --git a/gst/mpegtsdemux/mpegtspacketizer.c b/gst/mpegtsdemux/mpegtspacketizer.c
index 263f90f..d03afed 100644
--- a/gst/mpegtsdemux/mpegtspacketizer.c
+++ b/gst/mpegtsdemux/mpegtspacketizer.c
@@ -2284,7 +2284,7 @@ mpegts_packetizer_pts_to_ts (MpegTSPacketizer2 * packetizer,
     if (G_UNLIKELY (pcr_pid != 0x1fff &&
             ABSDIFF (res, pcrtable->last_pcrtime) > 15 * GST_SECOND))
       res = GST_CLOCK_TIME_NONE;
-    else {
+    else if GST_CLOCK_TIME_IS_VALID (pcrtable->base_pcrtime){
       GstClockTime tmp = pcrtable->base_time + pcrtable->skew;
       if (tmp + res >= pcrtable->base_pcrtime) {
         res += tmp - pcrtable->base_pcrtime;
@@ -2297,6 +2297,11 @@ mpegts_packetizer_pts_to_ts (MpegTSPacketizer2 * packetizer,
         res = GST_CLOCK_TIME_NONE;
       }
     }
+    else
+    {
+      res = GST_CLOCK_TIME_NONE;
+      GST_DEBUG ("set pts as -1, base_pcrtime:%" GST_TIME_FORMAT, GST_TIME_ARGS (pcrtable->base_pcrtime));
+    }
   } else if (packetizer->calculate_offset && pcrtable->groups) {
     gint64 refpcr = G_MAXINT64, refpcroffset;
     PCROffsetGroup *group = pcrtable->current->group;
