diff --git a/gst/mpegtsdemux/mpegtspacketizer.c b/gst/mpegtsdemux/mpegtspacketizer.c
index d03afed..02f9c00 100644
--- a/gst/mpegtsdemux/mpegtspacketizer.c
+++ b/gst/mpegtsdemux/mpegtspacketizer.c
@@ -1431,7 +1431,10 @@ calculate_skew (MpegTSPacketizer2 * packetizer,
          * to be resynched the next time we see a PCR */
         GST_WARNING
             ("backward timestamps at server or no buffer timestamps. Resync base PCR");
-        pcr->base_pcrtime = GST_CLOCK_TIME_NONE;
+        GST_DEBUG ("delta last_pcrtime - gstpcrtime:%" GST_TIME_FORMAT, GST_TIME_ARGS (pcr->last_pcrtime - gstpcrtime));
+        if (pcr->last_pcrtime - gstpcrtime > 2*GST_SECOND) {
+          pcr->base_pcrtime = GST_CLOCK_TIME_NONE;
+        }
       }
     }
   } else
