diff --git a/gst-libs/gst/codecparsers/gsth264parser.c b/gst-libs/gst/codecparsers/gsth264parser.c
index ec5cd3a..ce71039 100644
--- a/gst-libs/gst/codecparsers/gsth264parser.c
+++ b/gst-libs/gst/codecparsers/gsth264parser.c
@@ -2749,6 +2749,13 @@ gst_h264_video_calculate_framerate (const GstH264SPS * sps,
             case 3:
             case 4:
               delta_tfi_divisor = 2;
+              /*indicate variable frame rate,and if pic_struct equals GST_H264_SEI_PIC_STRUCT_TOP_BOTTOM
+              GST_H264_SEI_PIC_STRUCT_BOTTOM_TOP, it indicates interlace stream,so we double frame rate*/
+              if (!vui->fixed_frame_rate_flag) {
+                delta_tfi_divisor = 1;
+                GST_WARNING("variable frame rate,fixed_frame_rate_flag:%d, reset to delta_tfi_divisor:%d,double frame rate", 
+                  vui->fixed_frame_rate_flag,delta_tfi_divisor);
+              }
               break;
             case 5:
             case 6:
