diff --git a/bfd/dwarf2.c b/bfd/dwarf2.c
index dd3568a8532..8718b52d2c8 100644
--- a/bfd/dwarf2.c
+++ b/bfd/dwarf2.c
@@ -3248,7 +3248,7 @@ scan_unit_for_symbols (struct comp_unit *unit)
       else
 	{
 	  func = NULL;
-	  if (abbrev->tag == DW_TAG_variable)
+	  if (abbrev->tag == DW_TAG_variable || abbrev->tag == DW_TAG_member)
 	    {
 	      bfd_size_type amt = sizeof (struct varinfo);
 	      var = (struct varinfo *) bfd_zalloc (abfd, amt);
