diff --git a/hwdb.d/meson.build b/hwdb.d/meson.build
index e79380db24..450b37e5bd 100644
--- a/hwdb.d/meson.build
+++ b/hwdb.d/meson.build
@@ -6,17 +6,12 @@
 hwdb_files_notest = files(
         'README',
         '20-dmi-id.hwdb',
-        '20-pci-vendor-model.hwdb',
-        '20-pci-classes.hwdb',
         '20-usb-vendor-model.hwdb',
         '20-usb-classes.hwdb',
         '20-sdio-vendor-model.hwdb',
         '20-sdio-classes.hwdb',
         '20-bluetooth-vendor-product.hwdb',
-        '20-acpi-vendor.hwdb',
-        '20-OUI.hwdb',
-        '20-net-ifname.hwdb',
-        '20-vmbus-class.hwdb')
+        '20-net-ifname.hwdb')
 
 hwdb_files_test = files(
         '60-autosuspend.hwdb',
@@ -29,10 +24,7 @@ hwdb_files_test = files(
         '70-analyzers.hwdb',
         '70-av-production.hwdb',
         '70-cameras.hwdb',
-        '70-joystick.hwdb',
         '70-mouse.hwdb',
-        '70-pointingstick.hwdb',
-        '70-touchpad.hwdb',
         '80-ieee1394-unit-function.hwdb')
 
 if conf.get('ENABLE_HWDB') == 1
