Index: git/utils/bt-halplay.c
===================================================================
--- git.orig/utils/bt-halplay.c
+++ git/utils/bt-halplay.c
@@ -224,7 +224,7 @@ static void *pcm_worker_routine(void *ar
 		error("output speaker open failed\n");
 		goto fail;
 	}
-	snd_pcm_uframes_t period_size = 128;
+	snd_pcm_uframes_t period_size = 128*4;
 	ssize_t frame_size = 2 * w->transport.channels; //snd_pcm_frames_to_bytes(w->pcm, 1); //16bit 2channels
 	buffer = malloc(period_size * frame_size);
 	char *buffer_tail = buffer;
