Index: git/src/bluez-a2dp.c
===================================================================
--- git.orig/src/bluez-a2dp.c
+++ git/src/bluez-a2dp.c
@@ -36,6 +36,30 @@ const a2dp_sbc_t bluez_a2dp_sbc = {
 	.max_bitpool = MAX_BITPOOL,
 };
 
+const a2dp_sbc_t bluez_a2dp_src_fixed_48k_sbc = {
+	.frequency =
+		SBC_SAMPLING_FREQ_48000,
+	.channel_mode =
+		SBC_CHANNEL_MODE_MONO |
+		SBC_CHANNEL_MODE_DUAL_CHANNEL |
+		SBC_CHANNEL_MODE_STEREO |
+		SBC_CHANNEL_MODE_JOINT_STEREO,
+	.block_length =
+		SBC_BLOCK_LENGTH_4 |
+		SBC_BLOCK_LENGTH_8 |
+		SBC_BLOCK_LENGTH_12 |
+		SBC_BLOCK_LENGTH_16,
+	.subbands =
+		SBC_SUBBANDS_4 |
+		SBC_SUBBANDS_8,
+	.allocation_method =
+		SBC_ALLOCATION_SNR |
+		SBC_ALLOCATION_LOUDNESS,
+	.min_bitpool = MIN_BITPOOL,
+	.max_bitpool = MAX_BITPOOL,
+};
+
+
 #if ENABLE_MP3
 const a2dp_mpeg_t bluez_a2dp_mpeg = {
 	.layer =
Index: git/src/bluez-a2dp.h
===================================================================
--- git.orig/src/bluez-a2dp.h
+++ git/src/bluez-a2dp.h
@@ -18,6 +18,7 @@
 #include "a2dp-codecs.h"
 
 const a2dp_sbc_t bluez_a2dp_sbc;
+const a2dp_sbc_t bluez_a2dp_src_fixed_48k_sbc;
 #if ENABLE_MP3
 const a2dp_mpeg_t bluez_a2dp_mpeg;
 #endif
Index: git/src/bluez.c
===================================================================
--- git.orig/src/bluez.c
+++ git/src/bluez.c
@@ -746,7 +746,7 @@ void bluez_register_a2dp(void) {
 				A2DP_CODEC_VENDOR_APTX, &bluez_a2dp_aptx, sizeof(bluez_a2dp_aptx));
 #endif
 		bluez_register_a2dp_endpoint(BLUETOOTH_UUID_A2DP_SOURCE, BLUETOOTH_PROFILE_A2DP_SOURCE,
-				A2DP_CODEC_SBC, &bluez_a2dp_sbc, sizeof(bluez_a2dp_sbc));
+				A2DP_CODEC_SBC, &bluez_a2dp_src_fixed_48k_sbc, sizeof(bluez_a2dp_src_fixed_48k_sbc));
 	}
 	if (config.enable.a2dp_sink) {
 #if ENABLE_AAC
