diff --git a/src/connection.c b/src/connection.c
index d0c7d9f..e906afe 100644
--- a/src/connection.c
+++ b/src/connection.c
@@ -1012,9 +1012,12 @@ wl_closure_invoke(struct wl_closure *closure, uint32_t flags,
 
 	implementation = target->implementation;
 	if (!implementation[opcode]) {
-		wl_abort("listener function for opcode %u of %s is NULL\n",
+		wl_log("listener function for opcode %u of %s is NULL, return now\n",
 			 opcode, target->interface->name);
+		wl_closure_clear_fds(closure);
+		return;
 	}
+
 	ffi_call(&cif, implementation[opcode], NULL, ffi_args);
 
 	wl_closure_clear_fds(closure);
