From 1c476e379af1f8e12c4ba8ca56a2858a407e307d Mon Sep 17 00:00:00 2001
From: Blance Tang <guofeng.tang@amlogic.com>
Date: Thu, 10 Dec 2020 21:12:51 +0800
Subject: [PATCH] Always return true for wait

---
 scripts/make.sh | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/scripts/make.sh b/scripts/make.sh
index 77e0e06..3067d8e 100755
--- a/scripts/make.sh
+++ b/scripts/make.sh
@@ -312,7 +312,7 @@ do
   do
     [ "$COUNT" -lt "$CPUS" ] && break;
 
-    wait $j
+    wait $j || true
     DONE=$(($DONE+$?))
     COUNT=$(($COUNT-1))
     PENDING="${PENDING## $j}"
@@ -324,7 +324,7 @@ done
 
 for i in $PENDING
 do
-  wait $i
+  wait $i || true
   DONE=$(($DONE+$?))
 done
 
-- 
2.7.4

