diff --git a/gst/gstbufferpool.c b/gst/gstbufferpool.c
index 96192cf..66b9c26 100644
--- a/gst/gstbufferpool.c
+++ b/gst/gstbufferpool.c
@@ -544,7 +544,7 @@ gst_buffer_pool_set_active (GstBufferPool * pool, gboolean active)
      * freed when they are released */
     outstanding = g_atomic_int_get (&priv->outstanding);
     GST_LOG_OBJECT (pool, "outstanding buffers %d", outstanding);
-    if (outstanding == 0) {
+    if (outstanding >= 0) {
       if (!do_stop (pool))
         goto stop_failed;
     }
