diff --git a/gst/gstclock.c b/gst/gstclock.c
index 7f6b124..8a395a1 100644
--- a/gst/gstclock.c
+++ b/gst/gstclock.c
@@ -1010,7 +1010,11 @@ gst_clock_adjust_unlocked (GstClock * clock, GstClockTime internal)
       cnum, cdenom);

   /* make sure the time is increasing */
-  priv->last_time = MAX (ret, priv->last_time);
+  if(priv->last_time == GST_CLOCK_TIME_NONE){
+    priv->last_time = ret;
+  }else{
+    priv->last_time = MAX (ret, priv->last_time);
+  }

   return priv->last_time;
 }
