diff --git a/plugins/elements/gstmultiqueue.c b/plugins/elements/gstmultiqueue.c
index 8aa924c..a739c85 100644
--- a/plugins/elements/gstmultiqueue.c
+++ b/plugins/elements/gstmultiqueue.c
@@ -1494,8 +1494,9 @@ update_buffering (GstMultiQueue * mq, GstSingleQueue * sq)
     }

     if (is_buffering && buffering_level < mq->low_watermark) {
-      mq->buffering = TRUE;
-      SET_PERCENT (mq, percent);
+       // Do not post buffering because issue SWPL-78509
+       mq->buffering = FALSE;
+       // SET_PERCENT (mq, percent);
     }
   }
 }
