diff --git a/gst/gstprotection.c b/gst/gstprotection.c
index 73deed8..af6066b 100644
--- a/gst/gstprotection.c
+++ b/gst/gstprotection.c
@@ -93,7 +93,7 @@ gst_protection_meta_transform (GstBuffer * transbuf, GstMeta * meta,

   if (GST_META_TRANSFORM_IS_COPY (type)) {
     GstMetaTransformCopy *copy = data;
-    if (!copy->region) {
+    if (!copy->region || !gst_structure_has_name(protection_meta->info, "subsamples")) {
       /* only copy if the complete data is copied as well */
       gst_buffer_add_protection_meta (transbuf,
           gst_structure_copy (protection_meta->info));
