diff --git a/gst/typefind/gsttypefindfunctions.c b/gst/typefind/gsttypefindfunctions.c
index 5a4df31..1eefbaa 100644
--- a/gst/typefind/gsttypefindfunctions.c
+++ b/gst/typefind/gsttypefindfunctions.c
@@ -3497,6 +3497,12 @@ qt_type_find (GstTypeFind * tf, gpointer unused)
       }
     }
 
+    if (strncmp (&data[4], "ftypwmf ", 8) == 0) {
+        tip = GST_TYPE_FIND_MAXIMUM;
+        variant = "wmf";
+        break;
+    }
+
     /* top-level box/atom types that are in common with ISO base media file format */
     if (memcmp (&data[4], "moov", 4) == 0 ||
         memcmp (&data[4], "mdat", 4) == 0 ||
