diff --git a/gst/playback/gstplaybin2.c b/gst/playback/gstplaybin2.c
index c02ab72..b0de686 100644
--- a/gst/playback/gstplaybin2.c
+++ b/gst/playback/gstplaybin2.c
@@ -3612,6 +3612,10 @@ pad_added_cb (GstElement * decodebin, GstPad * pad, GstSourceGroup * group)
        * it will most likely stop. */
       if (combine->has_always_ok) {
         gboolean always_ok = (decodebin == group->suburidecodebin);
+        /*fix change multi audio track freeze for a few seconds*/
+        if (combine->type == GST_PLAY_SINK_TYPE_AUDIO && decodebin == group->uridecodebin) {
+          always_ok = TRUE;
+        }
         g_object_set (sinkpad, "always-ok", always_ok, NULL);
       }
       g_signal_emit (G_OBJECT (playbin), gst_play_bin_signals[signal], 0, NULL);
