diff --git a/gst/playback/gstplaysink.c b/gst/playback/gstplaysink.c
index 631ce7b..45be8cb 100644
--- a/gst/playback/gstplaysink.c
+++ b/gst/playback/gstplaysink.c
@@ -45,7 +45,8 @@ GST_DEBUG_CATEGORY_STATIC (gst_play_sink_debug);
 #define VOLUME_MAX_DOUBLE 10.0
 
 #define DEFAULT_FLAGS             GST_PLAY_FLAG_AUDIO | GST_PLAY_FLAG_VIDEO | GST_PLAY_FLAG_TEXT | \
-                                  GST_PLAY_FLAG_SOFT_VOLUME | GST_PLAY_FLAG_SOFT_COLORBALANCE
+                                  GST_PLAY_FLAG_SOFT_VOLUME | GST_PLAY_FLAG_SOFT_COLORBALANCE | \
+                                  GST_PLAY_FLAG_NATIVE_VIDEO
 
 #define GST_PLAY_CHAIN(c) ((GstPlayChain *)(c))
 
@@ -3925,8 +3926,10 @@ gst_play_sink_set_flags (GstPlaySink * playsink, GstPlayFlags flags)
 {
   g_return_val_if_fail (GST_IS_PLAY_SINK (playsink), FALSE);
 
+  // Add GST_PLAY_FLAG_NATIVE_VIDEO
+  // Because Amlogic UGMP Pipeline do not need video-convert and video-scale
   GST_OBJECT_LOCK (playsink);
-  playsink->flags = flags;
+  playsink->flags = flags | GST_PLAY_FLAG_NATIVE_VIDEO;
   GST_OBJECT_UNLOCK (playsink);
 
   return TRUE;
